WalletWasabi/Services/WalletService.cs

```diff
@@ -1137,101 +1137,80 @@ public Operation(Script script, Money amount, string label)
+			// Use the sanity check that under 2 satoshi per bytes should not be displayed. To correct possible rounding errors.
+			feePerBytes = FeeRate.Max(feePerBytes, new FeeRate(1m));
 
-			feePerBytes = feePerBytes.SatoshiPerByte < 1 ? new FeeRate(1m) : feePerBytes; // Use the sanity check that under 2 satoshi per bytes should not be displayed. To correct possible rounding errors.
-
-			bool spendAll = spendAllCount == 1;
-			int inNum;
-			if (spendAll)
-			{
-				inNum = allowedSmartCoinInputs.Count;
-			}
-			else
-			{
-				int expectedMinTxSize = 1 * Constants.P2wpkhInputSizeInBytes + 1 * Constants.OutputSizeInBytes + 10;
-				inNum = SelectCoinsToSpend(allowedSmartCoinInputs, toSend.Select(x => x.Amount).Sum() + feePerBytes.GetTotalFee(expectedMinTxSize)).Count();
-			}
-
-			// https://bitcoincore.org/en/segwit_wallet_dev/#transaction-fee-estimation
-			// https://bitcoin.stackexchange.com/a/46379/26859
-			int outNum = spendAll ? toSend.Length : toSend.Length + 1; // number of addresses to send + 1 for change
-			int vSize = NBitcoinHelpers.CalculateVsizeAssumeSegwit(inNum, outNum);
-			Logger.LogInfo<WalletService>($"Estimated tx size: {vSize} vbytes.");
-
-			// Multiply the standard by 1.1 to avoid possible off by one errors in flawed implementations.
-			var minFeeRate = new FeeRate(1.1m * new StandardTransactionPolicy().MinRelayTxFee.SatoshiPerByte);
-			if (feePerBytes < minFeeRate)
-			{
-				feePerBytes = minFeeRate;
-			}
-
-			Money fee = feePerBytes.GetTotalFee(vSize);
-
-			Logger.LogInfo<WalletService>($"Fee: {fee.Satoshi} Satoshi.");
+			var smartCoinsByOutpoint = allowedSmartCoinInputs.ToDictionary(s => s.GetOutPoint());
+			TransactionBuilder builder = Network.CreateTransactionBuilder();
+			builder.SetCoinSelector(new SmartCoinSelector(smartCoinsByOutpoint));
+			builder.AddCoins(allowedSmartCoinInputs.Select(c => c.GetCoin()));
 
-			// 5. How much to spend?
-			long toSendAmountSumInSatoshis = toSend.Select(x => x.Amount).Sum(); // Does it work if I simply go with Money class here? Is that copied by reference of value?
-			var realToSend = new (Script script, Money amount, string label)[toSend.Length];
 			for (int i = 0; i < toSend.Length; i++) // clone
 			{
-				realToSend[i] = (
-					new Script(toSend[i].Script.ToString()),
-					new Money(toSend[i].Amount.Satoshi),
-					toSend[i].Label);
-			}
-			for (int i = 0; i < realToSend.Length; i++)
-			{
-				if (realToSend[i].amount == Money.Zero) // means spend all
+				var destination = toSend[i].Script;
+				var amount = toSend[i].Amount;
+				var label = toSend[i].Label;
+				if (amount == Money.Zero)
 				{
-					realToSend[i].amount = allowedSmartCoinInputs.Select(x => x.Amount).Sum();
-
-					realToSend[i].amount -= new Money(toSendAmountSumInSatoshis);
-
-					if (subtractFeeFromAmountIndex is null)
-					{
-						realToSend[i].amount -= fee;
-					}
+					builder.SendAll(destination);
```

Comment:
When I call this method as follow:
```c#
var txResult = walletService.BuildTransaction(password, new[] { 
				new WalletService.Operation(scriptPubKey, Money.Zero, "foo") }, .....
```
I get the following error:

```
System.InvalidOperationException : No call to TransactionBuilder.Send has been done which can support the fees
``` 

@nopara73 is it okay to call it with an `Operation` with `amount == 0`? In which cases?

Reply:
Forget my silly question, yes, it is okay to call `Operation` with `Money.Zero`. I tested this same scenario with master and the problem is handled with this:
https://github.com/zkSNACKs/WalletWasabi/blob/02e148cc663a7d599ba326dc8f9faa5fdbc4ef0f/WalletWasabi/Services/WalletService.cs#L1040-L1043

This is a breaking change.

Reply:
A test case for this should be added as well

Reply:
I am on that.

Reply:
@lontivero the `BuildTransaction` should work, `No call to TransactionBuilder.Send has been done which can support the fees` is very strange. I will check your tests.

Reply:
OK I know why, I should have called `SubstractFees`. I will add a test to `SendTestsFromHiddenWalletAsync` to cover this case.